make/git_fetch.py

```diff
@@ -0,0 +1,218 @@
+
+    Returns:
+        A boolean value True if the hashes match of False if they don't
+    """
+    popen = subprocess.Popen(
+        ['sha256sum {}'.format(filename)], shell=True, stdout=subprocess.PIPE)
+    stdout, _ = popen.communicate()
+    if hash_file in os.listdir(os.getcwd()):
+        with open(hash_file, 'r+') as hashfp:
+            if hashfp.readline() == stdout.decode('utf-8'):
+                return True
+            else:
+                hashfp.truncate(0)
+                hashfp.write(stdout.decode('utf-8'))
+                hashfp.flush()
+    else:
+        with open(hash_file, 'w+') as hashfp:
+            hashfp.write(stdout.decode('utf-8'))
+            hashfp.flush()
+    return False
+
+
+def clean_up(filename):
+    """Cleans up by deleting the filename specified
+
+    Also deletes files that may have been downloaded extra and not cleaned up
+    properly.
+
+    Args:
+        filename: a string of the file to delete
+
+    Returns:
+        None
+    """
+    pattern = re.compile(filename + r'\.*[0-9]*')
+    for fname in os.listdir(os.getcwd()):
+        if re.search(pattern, fname):
+            os.remove(os.path.join(os.getcwd(), fname))
+
+
+def unpack(filename):
+    """Unpacks the specified zip file
+
+    Places the *.h files into inc/ and *.c into src/ and will create any missing
+    directories. It will also delete any previous files in these folers.
+
+    Args:
+        filename: string containing the name of the zip file to unpack
+
+    Returns:
+        Boolean value indicating success or failure
+    """
+    _, zip_ext = os.path.splitext(os.path.join(os.getcwd(), filename))
+    if zip_ext != '.zip':
+        return False
+
+    lsdir = os.listdir(os.getcwd())
+    with zipfile.ZipFile(os.path.join(os.getcwd(), filename), 'r') as zip_ref:
+        zip_ref.extractall(os.getcwd())
+    lsdir_post = os.listdir(os.getcwd())
+    newdir = list(set(lsdir_post) - set(lsdir))
+    if len(newdir) != 1:
+        return False
+
+    if 'inc' in lsdir:
+        clean_folder(os.path.join(os.getcwd(), 'inc'))
+    else:
+        os.mkdir(os.path.join(os.getcwd(), 'inc'))
+    if 'src' in lsdir:
+        clean_folder(os.path.join(os.getcwd(), 'src'))
+    else:
+        os.mkdir(os.path.join(os.getcwd(), 'src'))
+
+    for file_path in os.listdir(os.path.join(os.getcwd(), newdir[0])):
+        _, ext = os.path.splitext(file_path)
+        if ext == '.h':
+            os.rename(
+                os.path.join(os.getcwd(), newdir[0], file_path),
+                os.path.join(os.getcwd(), 'inc', os.path.basename(file_path)))
+        if ext == '.c':
+            os.rename(
+                os.path.join(os.getcwd(), newdir[0], file_path),
+                os.path.join(os.getcwd(), 'src', os.path.basename(file_path)))
+
+    # Delete the unzipped dir
+    shutil.rmtree(newdir[0])
+    return True
+
+
+def clean_folder(directory, ignore='tmp.c'):
```

Comment:
This seems to only be called by `unpack`. Is there a reason why `ignore` is a string and not just a list of files?

Reply:
Oh is this because dealing with Python default arguments are evaluated once when the function is defined so you might accidentally share state and need extra code to handle that?

```python
def clean_folder(directory, ignore=None):
    if ignore is None:
        ignore = []
    # ...
```

Reply:
Yes exactly that, it is considered dangerous to have a list as a default kwarg in a Python.